presto-hive/src/main/java/com/facebook/presto/hive/util/AsyncQueue.java

```diff
@@ -119,22 +134,104 @@ else if (elements.size() >= targetQueueSize) {
 
     public synchronized ListenableFuture<List<T>> getBatchAsync(int maxSize)
     {
+        return borrowBatchAsync(maxSize, elements -> new BorrowResult<>(ImmutableList.of(), elements));
+    }
+
+    /**
+     * Invoke {@code function} with up to {@code maxSize} elements removed from the head of the queue,
+     * and insert elements in the return value to the tail of the queue.
+     * <p>
+     * If no element is currently available, invocation of {@code function} will be deferred until some
+     * element is available, or no more elements will be. Spurious invocation of {@code function} is
+     * possible.
+     * <p>
+     * Insertion through return value of {@code function} will be effective even if {@link #finish()} has been invoked.
+     * When borrow (of a non-empty list) is ongoing, {@link #isFinished()} will return false.
+     * If an empty list is supplied to {@code function}, it must not return a result indicating intention
+     * to insert elements into the queue.
+     */
+    public <O> ListenableFuture<O> borrowBatchAsync(int maxSize, Function<List<T>, BorrowResult<T, O>> function)
+    {
         checkArgument(maxSize >= 0, "maxSize must be at least 0");
 
-        List<T> list = getBatch(maxSize);
-        if (!list.isEmpty()) {
-            return immediateFuture(list);
-        }
-        else if (finishing) {
-            return immediateFuture(ImmutableList.of());
-        }
-        else {
-            return Futures.transform(notEmptySignal, x -> getBatch(maxSize), executor);
+        ListenableFuture<List<T>> borrowedListFuture;
+        AtomicBoolean borrowerIncremented = new AtomicBoolean();
```

Comment:
I don't think you need this.  You only increment when the list is not empty so you can just check that in the transform below